internal/tools/aws/database_perseus.go

```diff
@@ -0,0 +1,1251 @@
+		context.TODO(),
+		&secretsmanager.GetSecretValueInput{
+			SecretId: rdsCluster.DBClusterIdentifier,
+		})
+	if err != nil {
+		return errors.Wrapf(err, "failed to find the master secret for the multitenant proxy cluster %s", rdsID)
+	}
+
+	close, err := d.connectToRDSCluster(rdsPostgresDefaultSchema, *rdsCluster.Endpoint, DefaultMattermostDatabaseUsername, *masterSecretValue.SecretString)
+	if err != nil {
+		return errors.Wrapf(err, "failed to connect to the multitenant proxy cluster %s", rdsID)
+	}
+	defer close(logger)
+
+	ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(DefaultMySQLContextTimeSeconds*time.Second))
+	defer cancel()
+
+	err = d.ensureDatabaseIsCreated(ctx, databaseName)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create database in multitenant proxy cluster %s", rdsID)
+	}
+
+	return nil
+}
+
+func (d *PerseusDatabase) ensureLogicalDatabaseSetup(databaseName, vpcID string, rdsCluster *rdsTypes.DBCluster, logger log.FieldLogger) (*RDSSecret, error) {
+	rdsID := *rdsCluster.DBClusterIdentifier
+
+	masterSecretValue, err := d.client.Service().secretsManager.GetSecretValue(
+		context.TODO(),
+		&secretsmanager.GetSecretValueInput{
+			SecretId: rdsCluster.DBClusterIdentifier,
+		})
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to find the master secret for the multitenant proxy cluster %s", rdsID)
+	}
+
+	close, err := d.connectToRDSCluster(databaseName, *rdsCluster.Endpoint, DefaultMattermostDatabaseUsername, *masterSecretValue.SecretString)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to connect to the multitenant proxy cluster %s", rdsID)
+	}
+	defer close(logger)
+
+	ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(DefaultMySQLContextTimeSeconds*time.Second))
+	defer cancel()
+
+	err = d.ensureBaselineDatabasePrep(ctx, databaseName)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to perform basic database setup")
+	}
+
+	// Run PGBouncer prep to provide a clean path for migrations to and from
+	// Perseus.
+	err = d.ensurePGBouncerDatabasePrep(ctx, databaseName, DefaultPGBouncerAuthUsername)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to perform pgbouncer setup")
+	}
+
+	installationSecret, err := d.ensureMultitenantDatabaseSecretIsCreated(rdsCluster.DBClusterIdentifier, &vpcID)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get secret for installation")
+	}
+
+	err = d.ensureDatabaseUserIsCreated(ctx, installationSecret.MasterUsername, installationSecret.MasterPassword)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create Mattermost database user")
+	}
+
+	err = d.ensurePerseusUserHasRole(ctx, installationSecret.MasterUsername)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to grant permissions to perseus database user")
+	}
+
+	err = d.ensureMasterUserHasRole(ctx, DefaultPerseusDatabaseUsername)
```

Comment:
Argument is role, but we're sending username. Assuming same usrname and role names, but its confusing from this point of view.

I also find confusing how we create "DatabaseUser", "PerseusUser" and "MasterUser". Are all those three different? I'm asking because we are using the same values/db for some of them.

Reply:
Yeah, I know this part is hard to follow. There are indeed 4 users: master DB user, perseus user to connect to the auth database, perseus user to connect to the shared DBs for connection pooling, and the individual schemas+users for each Mattermost installation. I have plans to clean this up a bunch for both perseus and pgbouncer in a follow-up PR.